src/security/KeyData.cc

```diff
@@ -86,51 +88,44 @@ Security::KeyData::loadX509ChainFromFile()
-        CertPointer latestCert = cert;
-
-        while (const auto ca = Ssl::ReadX509Certificate(bio)) {
-            // get Issuer name of the cert for debug display
-            char *nameStr = X509_NAME_oneline(X509_get_subject_name(ca.get()), nullptr, 0);
-
-#if TLS_CHAIN_NO_SELFSIGNED // ignore self-signed certs in the chain
-            // self-signed certificates are not valid in a sent chain
-            if (X509_check_issued(ca.get(), ca.get()) == X509_V_OK) {
-                debugs(83, DBG_PARSE_NOTE(2), "CA " << nameStr << " is self-signed, will not be chained: " << nameStr);
-                OPENSSL_free(nameStr);
-                continue;
-            }
-#endif
-            // checks that the chained certs are actually part of a chain for validating cert
-            const auto checkCode = X509_check_issued(ca.get(), latestCert.get());
-            if (checkCode == X509_V_OK) {
-                debugs(83, DBG_PARSE_NOTE(3), "Adding issuer CA: " << nameStr);
-                // OpenSSL API requires that we order certificates such that the
-                // chain can be appended directly into the on-wire traffic.
-                latestCert = CertPointer(ca);
-                chain.emplace_front(latestCert);
-            } else {
-                debugs(83, DBG_PARSE_NOTE(2), certFile << ": Ignoring non-issuer CA " << nameStr << ": " << X509_verify_cert_error_string(checkCode) << " (" << checkCode << ")");
+    debugs(83, DBG_PARSE_NOTE(3), "Using certificate chain in " << certFile);
+    // and add to the chain any other certificate exist in the file
+    CertList certsInFile;
+
+    while (const auto ca = Ssl::ReadX509Certificate(bio)) {
+        certsInFile.emplace_back(ca);
+    }
+
+    CertPointer latestCert = cert;
+    do {
+        CertPointer anIssuer;
+        // get the cert name for debug display
+        char *nameStr = X509_NAME_oneline(X509_get_subject_name(latestCert.get()), nullptr, 0);
+        // self-signed certificate, probably the root certificate, must not chained.
+        if (X509_check_issued(latestCert.get(), latestCert.get()) == X509_V_OK) {
+            debugs(83, DBG_PARSE_NOTE(2), "CA " << nameStr << " is self-signed, will not be chained: " << nameStr);
+        } else {
+            for (auto candidateIssuer: certsInFile) {
+                const auto checkCode = X509_check_issued(candidateIssuer.get(), latestCert.get());
+                if (checkCode == X509_V_OK) {
+                    // We found an issuer add it to the chain.
+                    debugs(83, DBG_PARSE_NOTE(3), "Adding issuer CA: " << nameStr);
+                    anIssuer = CertPointer(candidateIssuer);
+                    chain.emplace_front(anIssuer);
+                    break;
+                }
             }
-            OPENSSL_free(nameStr);
         }
-    }
+        latestCert = anIssuer;
+        OPENSSL_free(nameStr);
+    } while (latestCert != nullptr);
```

Comment:
It looks like we are _not_ chaining the "top" issuer (i.e. certificate A in the "A issued B issued C" chain), even if it is not self-signed. Is that intentional?

Reply:
Certificate chaining in this loop are guarded by the X509_check_issued() check. Certificate A in the "A issued B issued C" chain will never pass that check because none of the certificates in the bundle has issued A. Thus, A will not be added, even if it is not self-signed. Am I missing something?

Reply:
The loop does the following:

```
//cert is the server certificate
latestCert = cert;
do {
     if (anIssuer=FindIssuerOf(latestCert)) {
           addCertInChain(anIssuer);
    }
   latestCert = anIssuer;
} while(latestCert);
```
The steps:
-  'A' is the server certificate
-  check if one of the certificates in list issued 'A'. This is the 'B'. 
 - Add the B to the (end of a the empty) chain. 
 - Check if any of the certificates in chain issued 'B'. This is the 'C'.
 - Add C to the end of chain.
 - Check if any of the certificates in list issued 'C'. None found.
 - stop here

But as I can see now there is a bug in code. The chain list is in the reverse order. I can not understand why the `chain.emplace_front(anIssuer)` is used here instead of the ``chain.emplace_back(anIssuer)``.
And this bug looks very old.
I fix this one in my last commit  413f4777ac
Does this clarify?

Reply:
I figured it out. Our debugs() says "Adding issuer CA: latestCert", but the code actually adds candidateIssuer rather than the latestCert used in that debugs(). Thus, the loop is skipping the _leaf_ certificate (as expected) rather than the top issuer certificate (as I incorrectly thought). Resolving.

I may polish this debugs() to say "adding issuer CA for latestCert" or "adding issuer of latestCert" or something like that.

----

>  I can not understand why the chain.emplace_front(anIssuer) is used here instead of the chain.emplace_back(anIssuer).

I agree with your bug assessment and your fix. OpenSSL says that certificates added via SSL_CTX_add_extra_chain_cert() are sent "in order" (presumably in that function call order) so, based on your RFC 8446 Section 4.4.2 quote, we should maintain leaf-to-root order in the chain, adding the (top) issuer to the _back_.

It is interesting (or even strange) that this bug had no known effects on agents Squid interoperates with.